examples/gno.land/p/demo/subscription/subscription.gno

```diff
@@ -0,0 +1,123 @@
+//
+//		Perform payment for a lifetime subscription.
+//		lifetimeSub.ProcessPayment(caller, 500)
+//	}
+
+package subscription // import "gno.land/p/demo/subscription"
+
+import (
+	"gno.land/p/demo/ufmt"
+	"gno.land/p/demo/avl"
+	"time"
+	"std"
+)
+
+// Subscription interface defines standard methods that all subscription types must implement.
+type Subscription interface {
+	CheckAccess(std.Address)
+	ProcessPayment(std.Address, int64)
+}
+
+// RecurringSubscription represents a subscription that requires periodic payments.
+// It includes the duration of the subscription and the amount required per period.
+type RecurringSubscription struct {
+	duration time.Duration 
+	amount   int64
+	subs     *avl.Tree      // std.Address -> time.Time
+}
+
+// LifetimeSubscription represents a subscription that requires only a one-time payment.
+// It grants permanent access to a service or product.
+type LifetimeSubscription struct {
+	amount int64
+	subs   *avl.Tree    // std.Address -> bool
+}
+
+// NewRecurringSubscription creates and returns a new recurring subscription.
+func NewRecurringSubscription(duration time.Duration, amount int64) *RecurringSubscription {
+	return &RecurringSubscription{
+		duration: duration,
+		amount:   amount,
+		subs:     avl.NewTree(),
+	}
+}
+
+// CheckAccess verifies if the caller has an active recurring subscription.
+func (rs *RecurringSubscription) CheckAccess(caller std.Address) {
+	send := std.GetOrigSend()
+	sendAmount := send.AmountOf("ugnot")
+
+	if sendAmount < rs.amount {
+		panic(ufmt.Sprintf("you need to send at least %d units to access this feature", rs.amount))
+	}
+
+	expTime, exists := rs.subs.Get(caller.String())
+	if !exists || time.Now().After(expTime.(time.Time)) {
+		panic("your subscription has expired or does not exist")
+	}
+}
+
+// ProcessPayment processes the payment for a recurring subscription and extends its validity.
+func (rs *RecurringSubscription) ProcessPayment(caller std.Address, paymentAmount int64) {
+	if paymentAmount < rs.amount {
+		panic("insufficient payment")
+	}
+
+	expiration := time.Now().Add(rs.duration)
+	rs.subs.Set(caller.String(), expiration)
+}
+
+// GetExpiration returns the expiration date of the recurring subscription for a given caller.
+func (rs *RecurringSubscription) GetExpiration(caller std.Address) time.Time {
+	expTime, exists := rs.subs.Get(caller.String())
+	if !exists {
+		panic("no subscription found")
+	}
+
+	return expTime.(time.Time)
+}
+
+// NewLifetimeSubscription creates and returns a new lifetime subscription.
+func NewLifetimeSubscription(amount int64) *LifetimeSubscription {
+	return &LifetimeSubscription{
+		amount: amount,
+		subs:   avl.NewTree(),
+	}
+}
+
+// ProcessPayment processes the payment for a lifetime subscription.
+func (ls *LifetimeSubscription) ProcessPayment(caller std.Address, paymentAmount int64) {
+	if paymentAmount < ls.amount {
+		panic(ufmt.Sprintf("insufficient payment, required %d units", ls.amount))
```

Comment:
For `p/`, it's better to return an error instead of panicking. However, you can then create a "MustProcessPayment" helper, where "Must" is the version that panics.

See https://docs.gno.land/concepts/effective-gno/#embrace-panic